lms/templates/courseware/courseware.html

```diff
@@ -112,6 +112,47 @@
       </script>
   % endif
 
+  <script type="text/javascript">
+    /* Helper function isInViewport checks whether
+    the given element is in viewport vertically or not */
+    function isInViewport(el) {
+      const scroll = window.scrollY || window.pageYOffset
+      const elementTop = el.getBoundingClientRect().top + scroll
+      const viewport = {
+        top: scroll,
+        bottom: scroll + window.innerHeight,
+      }
+      const elementMid = elementTop + el.clientHeight/2
+      // Returns true if the middle of the element is in the viewport.
+      return elementMid >= viewport.top && elementMid <= viewport.bottom
+    }
+
+    /*Add a jQuery plugin to override the focus behavior.
+    When focused, the element will be the scrolled to bottom of the viewport.
+    */
+    (function ($) {
```

Comment:
Do you have a sense of how often this `focus` function is called?  It seems like we're overriding it for all calls on this page... I'd just want to sanity check that we're not adding a whole bunch of processing to the page when we loop through `this.each` below.

Reply:
@davidjoy  Thanks for the review and suggestion. I am working on it this week.

> Do you have a sense of how often this focus function is called?

I will try to get some metric on this and update you by end of the week.

Reply:
@davidjoy Following is the frequency of `focus` function calls:

1. `focus` is called `twice` when rendering the sequence i.e by [Sequence.render](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/js/src/sequence/display.js#L230). It is called on first render of page as well as whenever the active sequence is changed by clicking on `Next`, `Previous` or the sequence button.
2. `focus` is called `twice` for each Problem in the xblock when rendering. So, if the sequence has `n` problems, `focus` will be called `n*2` times.
3. `focus` is called `once` when rendering `discussions` on the sequence (xblock).
4. `focus` is called `twice` for videos in the sequence (xblock)
5. For staff users, `focus` is called once when rendering the staff toolbar on top of the page.

Additionally, `focus` is called `twice` by [calculator.js](https://github.com/edx/edx-platform/blob/master/lms/static/js/calculator.js) on first render of the page.

Please let me know you further thoughts.

---

I tried attaching a `delegated` ([ref](https://api.jquery.com/on/)) event handler on `.notification-btn` but the browser behaviour was preceded by the event handler callback. The notification was already moved to centre of page before the callback function ran.